system-tests/lib/benchspy/markdown_reporter.go

```diff
@@ -0,0 +1,413 @@
+package benchspy
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// MetricConfig represents configuration for a single metric
+type MetricConfig struct {
+	Group       string  // Which group this metric belongs to
+	UnitLabel   string  // Unit label (e.g. "ms", "MB")
+	ScaleFactor float64 // Scale factor for display (e.g. 1000.0 to convert ms to s)
+}
+
+type report struct {
+	TestName           string    `json:"test_name"`
+	CommitOrTag        string    `json:"commit_or_tag"`
+	TestStartTimestamp time.Time `json:"test_start_timestamp"`
+	TestEndTimestamp   time.Time `json:"test_end_timestamp"`
+	QueryExecutors     []struct {
+		Kind         string                 `json:"kind"`
+		QueryResults map[string]interface{} `json:"query_results"`
+	} `json:"query_executors"`
+}
+
+func bar(value, max, width float64, symbol string) string {
+	filled := value * width / max
+	return strings.Repeat(symbol, int(filled)) + strings.Repeat(" ", int(width-filled))
+}
+
+func createBarChart(
+	title string,
+	dates []time.Time,
+	values []float64,
+	maxValue float64, // If <= 0, calculated from values with 20% padding
+	unitLabel string,
+	scaleFactor float64, // e.g., 1000.0 to convert ms to s
+) string {
+	// Calculate max if not provided
+	if maxValue <= 0 {
+		for _, value := range values {
+			if value > maxValue {
+				maxValue = value
+			}
+		}
+		maxValue *= 1.2 // Add 20% padding
+	}
+
+	var b strings.Builder
+	b.WriteString(fmt.Sprintf("\n## %s\n\n", title))
+	b.WriteString("| Date       | Value |\n")
+	b.WriteString("| --- | --- |\n")
+
+	for i := 0; i < len(dates) && i < len(values); i++ {
+		scaledValue := values[i] / scaleFactor
```

Comment:
make sure `scaleFactor` is never `0`?

Reply:
Fixed in [`22a4315` (#18251)](https://github.com/smartcontractkit/chainlink/pull/18251/commits/22a4315c620c4d36b465e34aff1ab7b1926cf3db)